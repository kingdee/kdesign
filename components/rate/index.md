---
category: Components
type: 表单
order: 10
title: Rate
subtitle: 评分
---

单选是一种在一组互斥的选项中进行选择的组件。

## 使用场景

需要在一组互斥的选项进行选择时使用。

## API

| 属性 | 说明 | 类型 | 默认值 | 版本 |
| --- | --- | --- | --- | --- | --- |
| activeIcon | 自定义选中时的图标 | `React.ReactNode | ((props: unknown) => React.ReactNode)` | `<Icon type="star">` | 1.0.0 |
| allowHalf | 是否允许半选 | `true` `false` | `true` | 1.0.0 |
| color | 选中时 icon 颜色 | string | `-` | 1.0.0 |
| className | 自定义样式类名 | string | `-` | 1.0.0 |
| count | star 总数 | number | `5` | 1.0.0 |
| defaultValue | 默认选中的星星数 | number | `0` | 1.0.0 |
| disabled | 只读，无法选择 | `true` `false` | `false` | 1.0.0 |
| icon | 自定义显示的 icon | `ReactNode|(RateProps) => ReactNode` |`<Icon type="star">` | 1.0.0 |
| onlyActiveCurrent | 仅选中当前图标 | `true` `false`  | `false` | 1.0.0 |
| size | 评分尺寸（large/middle/small） | `large` `middle` `small` | `middle`  | 1.0.0 |
| style | 自定义样式 | CSSProperties | `-` |  1.0.0 |
| value | 当前选中的星星数，受控值 | number | `0` |  1.0.0 |
| onBlur | 失去焦点时的回调 | function() | `-` |  1.0.0 |
| onChange | 选择时的回调 | function(value: number) | `-` | 1.0.0 |
| onHoverChange | 鼠标 hover 时回调 | function(value: number) | `-` |1.0.0 |
| onFocus | 获取焦点时的回调 | function() | `-` | 1.0.0 |

## Design Token

| 分类    | 组件 token                          | 全局 token      | 默认值  |
| ------- | ----------------------------------- | --------------- | ------- |
| color   | --kd-c-rate-color-text-selected     | -               | #FEC104 |
|         | --kd-c-rate-color-text-not-selected | -               | #E5E5E5 |
| font    | --kd-c-rate-font-size-small         | -               | 12px    |
|         | --kd-c-rate-font-size-middle        | -               | 16px    |
|         | --kd-c-rate-font-size-large         | -               | 20px    |
| motion  | --kd-c-rate-motion-duration         | --kd-g-duration | 0.3s    |
| spacing | --kd-c-rate-spacing-margin-right    | -               | 8px     |
