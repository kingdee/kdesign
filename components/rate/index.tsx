import Rate from './rate'
export { RateProps as IRateProps } from './rate'
export default Rate
