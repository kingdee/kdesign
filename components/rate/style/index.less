@import '../../style/themes/index';
@import '../../style/mixins/index';
@import './token.less';

@rate-prefix-cls: ~'@{kd-prefix}-rate';
@rate-icon-prefix-cls: ~'@{rate-prefix-cls}-icon';
.@{rate-prefix-cls} {
  .reset-component;
  display: inline-flex;
  justify-content: flex-start;
  &-size-large {
    font-size: @rate-large-font-size;
  }
  &-size-middle {
    font-size: @rate-middle-font-size;
  }
  &-size-small {
    font-size: @rate-small-font-size;
  }
}
.@{rate-prefix-cls}-item {
  display: inline-flex;
  height: 100%;
  justify-content: center;
  align-items: center;
  line-height: 100%;
  margin-right: @rate-icon-margin;
  .@{rate-icon-prefix-cls} {
    &-view {
      position: relative;
      display: inline-flex;
      transition: transform @rate-transition-duration;
    }
    &-view:hover {
      transform: scale(1.2);
    }
    &-wrapper {
      display: inline-flex;
    }
    &-first {
      display: inline-flex;
      position: absolute;
      top: 0;
      left: 0;
      width: 50%;
      overflow: hidden;
      color: @rate-not-selected-color;
      opacity: 1;
      cursor: pointer;
      transition: color @rate-transition-duration;
    }
    &-first:hover {
      color: @rate-selected-color;
    }
    &-second {
      display: inline-flex;
      color: @rate-not-selected-color;
      opacity: 1;
      cursor: pointer;
      transition: color @rate-transition-duration;
      &-content {
        display: inline-flex;
      }
    }
    &-second:hover {
      color: @rate-selected-color;
    }
    &-selected {
      color: @rate-selected-color;
      opacity: 1;
    }
  }
}
.@{rate-prefix-cls}-item:last-child {
  margin-right: 0;
}

.@{rate-prefix-cls}-disabled {
  pointer-events: none;
}
