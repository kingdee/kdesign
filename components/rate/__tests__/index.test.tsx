import React from 'react'
import { render, mount } from 'enzyme'
import Icon from '../../icon'
import ConfigProvider from '../../config-provider/index'
// import ConfigProvider from '../../config-provider/index'
import Rate from '../index'
import { SizeTypes, size as SizeType } from '../rate'
import mountTest from '../../../tests/shared/mountTest'
const props = {
  onBlur: jest.fn(),
  onFocus: jest.fn(),
  onChange: jest.fn(),
  onHoverChange: jest.fn(),
}
describe('Rate', () => {
  // 1.mount test
  describe('1.mount test', () => {
    mountTest(Rate)
    SizeTypes.forEach((size) => {
      mountTest(() => <Rate size={size} />)
    })
  })

  // 2.render test
  describe('2.render test', () => {
    it('renders correctly', () => {
      expect(render(<Rate></Rate>)).toMatchSnapshot()
      SizeTypes.forEach((size) => {
        expect(render(<Rate size={size}></Rate>)).toMatchSnapshot()
      })
    })
  })

  // 3.warns in component
  describe('3.warns in component', () => {
    it('warns if size is wrong', () => {
      const mockWarn = jest.fn()
      jest.spyOn(console, 'warn').mockImplementation(mockWarn)
      const size = 'super' as any as SizeType
      render(<Rate size={size} />)
      expect(mockWarn).toHaveBeenCalledTimes(1)
      expect(mockWarn.mock.calls[0][0]).toMatch("Warning: [kdesign]-Rate: cannot found SizeType 'super'")
    })
  })

  // 4. render null or undefined without errors
  describe('4. render null or undefined without errors', () => {
    it('render null or undefined buttons without errors', () => {
      expect(
        mount(
          <Rate>
            {null}
            {undefined}
          </Rate>,
        ),
      ).toMatchSnapshot()
    })
  })

  // 5. displayName
  describe('5. displayName', () => {
    it('should have displayName static property', () => {
      const wrapper = mount(<Rate></Rate>)
      expect((wrapper.type() as any).displayName).toBe('Rate')
    })
  })

  // 6. class state
  describe('6.class state', () => {
    it('default star count is five', () => {
      const wrapper = mount(<Rate></Rate>)
      expect(wrapper.find('.kd-rate-item').length).toBe(5)
    })

    it('default icon is star and allow half choose', () => {
      const wrapper = mount(<Rate></Rate>)
      expect(wrapper.find('.kdicon-star').length).toBe(10)
    })

    it('allow half choose', () => {
      const wrapper = mount(<Rate></Rate>)
      expect(wrapper.find('.kd-rate-icon-first').length).toBe(5)
    })

    it('not allow half choose', () => {
      const wrapper = mount(<Rate allowHalf={false}></Rate>)
      expect(wrapper.find('.kd-rate-icon-first').length).toBe(0)
    })

    it('set rate read only', () => {
      const wrapper = mount(<Rate disabled={true}></Rate>)
      expect(wrapper.find('.kd-rate-disabled').length).toBe(1)
    })

    it('set rate default value which type is int', () => {
      const wrapper = mount(<Rate defaultValue={3}></Rate>)
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(6)
    })

    it('set rate default value which type is decimal and between 3.0 and 3.5, should not count half of a star', () => {
      const wrapper = mount(<Rate value={3.0}></Rate>)
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(6)
      wrapper.setProps({ value: 3.001 })
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(6)
      wrapper.setProps({ value: 3.499 })
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(6)
    })

    it('set rate default value which type is decimal and between 3.5 and 3.9 or equal 3.5, should count half of a star', () => {
      const wrapper = mount(<Rate value={3.5}></Rate>)
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(7)
      wrapper.setProps({ value: 3.999 })
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(7)
    })

    it('set rate count', () => {
      const wrapper = mount(<Rate count={10}></Rate>)
      expect(wrapper.find('.kd-rate-item').length).toBe(10)
    })

    it('set rate choosed color', () => {
      const wrapper = mount(<Rate color="red" defaultValue={3}></Rate>)
      const dom = wrapper.find('.kd-rate-icon-selected').at(0).getDOMNode() as HTMLElement
      expect(dom.style.color).toBe('red')
    })

    it('render different size', () => {
      const wrapper = mount(<Rate size={'small'}></Rate>)
      expect(wrapper.find('.kd-rate-size-small').length).toBe(1)
      wrapper.setProps({ size: 'middle' })
      expect(wrapper.find('.kd-rate-size-middle').length).toBe(1)
      wrapper.setProps({ size: 'large' })
      expect(wrapper.find('.kd-rate-size-large').length).toBe(1)
    })

    it('set rate only active current star', () => {
      const wrapper = mount(<Rate onlyActiveCurrent={true} defaultValue={3}></Rate>)
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(2)
    })
    it('replace rate active icon', () => {
      const _activeIcon = [
        <Icon type="search" key={'search_0'} />,
        <Icon type="search" key={'search_1'} />,
        <Icon type="search" key={'search_2'} />,
        <Icon type="search" key={'search_3'} />,
        <Icon type="search" key={'search_4'} />,
      ]
      const wrapper = mount(
        <Rate
          activeIcon={(index: number) => {
            return _activeIcon[index - 1]
          }}
          defaultValue={0.5}
        ></Rate>,
      )
      expect(wrapper.find('.kdicon-search').length).toBe(1)
    })

    it('set rate custom icon that is character', () => {
      const _icon = ['A', 'B', 'C', 'D', 'E']
      const wrapper = mount(
        <Rate
          allowHalf={false}
          icon={(index) => {
            return _icon[index - 1]
          }}
        ></Rate>,
      )
      expect(wrapper.find('.kd-rate-icon-wrapper').at(0).text()).toBe('A')
      expect(wrapper.find('.kd-rate-icon-wrapper').at(1).text()).toBe('B')
      expect(wrapper.find('.kd-rate-icon-wrapper').at(2).text()).toBe('C')
      expect(wrapper.find('.kd-rate-icon-wrapper').at(3).text()).toBe('D')
      expect(wrapper.find('.kd-rate-icon-wrapper').at(4).text()).toBe('E')
    })

    it('set rate custom icon that is react component', () => {
      const wrapper = mount(<Rate icon={<Icon type="warning-solid" />}></Rate>)
      expect(wrapper.find('.kdicon-warning-solid').length).toBe(10)
    })

    it('set rate custom style', () => {
      const wrapper = mount(<Rate style={{ fontSize: 36, backgroundColor: 'gray' }} defaultValue={0.5}></Rate>)
      const dom = wrapper.find('.kd-rate').at(0).getDOMNode() as HTMLElement
      expect(dom.style.fontSize).toBe('36px')
      expect(dom.style.backgroundColor).toBe('gray')
    })

    it('set rate custom className', () => {
      const className = 'kd-custom-className'
      const wrapper = mount(<Rate className={className}></Rate>)
      expect(wrapper.find('.kd-custom-className').length).toBe(2)
    })
  })
  // 7.component interaction(event)
  describe('7.component interaction(event)', () => {
    it('rate onBlur onFocus callback', () => {
      const wrapper = mount(<Rate {...props}></Rate>)
      wrapper.find('span').at(0).simulate('mouseenter')
      wrapper.find('span').at(0).simulate('mouseleave')
      expect(props.onFocus).toHaveBeenCalledTimes(1)
      expect(props.onBlur).toHaveBeenCalledTimes(1)
    })

    it('second star onHoverChange callback', () => {
      const wrapper = mount(<Rate {...props}></Rate>)
      wrapper.find('.kd-rate-icon-second').at(0).simulate('mouseenter')
      wrapper.find('.kd-rate-icon-second').at(0).simulate('mouseleave')
      expect(props.onHoverChange).toHaveBeenCalledTimes(1)
      props.onHoverChange.mockRestore()
    })

    it('second star onChange callback', () => {
      const wrapper = mount(<Rate {...props}></Rate>)
      wrapper.find('.kd-rate-icon-second').at(0).simulate('click')
      expect(props.onChange).toHaveBeenCalledTimes(1)
      props.onChange.mockRestore()
    })

    it('first star onHoverChange callback', () => {
      const wrapper = mount(<Rate {...props}></Rate>)
      wrapper.find('.kd-rate-icon-first').at(0).simulate('mouseenter')
      wrapper.find('.kd-rate-icon-first').at(0).simulate('mouseleave')
      expect(props.onHoverChange).toHaveBeenCalledTimes(1)
      props.onHoverChange.mockRestore()
    })

    it('first star onChange callback', () => {
      const wrapper = mount(<Rate {...props}></Rate>)
      wrapper.find('.kd-rate-icon-first').at(0).simulate('click')
      expect(props.onChange).toHaveBeenCalledTimes(1)
      props.onChange.mockRestore()
    })

    it('mouse hover or click star , then will active all left icon', () => {
      const wrapper = mount(<Rate {...props} defaultValue={3}></Rate>)
      wrapper.find('.kd-rate-icon-first').at(4).simulate('mouseenter')
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(9)
      wrapper.find('.kd-rate-icon-first').at(4).simulate('click')
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(9)

      wrapper.find('.kd-rate-icon-second').at(2).simulate('mouseenter')
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(6)
      wrapper.find('.kd-rate-icon-second').at(2).simulate('click')
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(6)
    })

    it('mouse hover or click star ,but active only current icon when set onlyActiveCurrent attribution to be true', () => {
      const wrapper = mount(<Rate {...props} defaultValue={3} onlyActiveCurrent></Rate>)
      wrapper.find('.kd-rate-icon-first').at(3).simulate('click')
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(1)
      wrapper.find('.kd-rate-icon-second').at(4).simulate('click')
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(2)

      wrapper.find('.kd-rate-icon-first').at(3).simulate('mouseenter')
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(1)
      wrapper.find('.kd-rate-icon-second').at(4).simulate('mouseenter')
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(2)
    })
  })

  // 8.config provider
  describe('8.config provider', () => {
    it('should config use config provider', () => {
      const rateConfig = {
        compDefaultProps: {
          Rate: {
            defaultValue: 3,
          },
        },
      }
      const wrapper = mount(
        <ConfigProvider value={rateConfig}>
          <Rate />
        </ConfigProvider>,
      )
      expect(wrapper.find('.kd-rate-icon-selected').length).toBe(6)
    })
  })

  // 9. ref test
  // 10. api test
  describe('10.api test', () => {
    it('API test', () => {
      // defaultValue
      const defaultValueRate = mount(<Rate defaultValue={1}></Rate>)
      expect(defaultValueRate.find('.kd-rate-icon-selected').length).toBe(2)
      // value
      const valueRate = mount(<Rate value={3.0}></Rate>)
      expect(valueRate.find('.kd-rate-icon-selected').length).toBe(6)
    })
  })
})
