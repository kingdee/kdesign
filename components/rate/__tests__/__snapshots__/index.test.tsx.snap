// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rate 2.render test renders correctly 1`] = `
<span
  class="kd-rate kd-rate-size-middle"
>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
</span>
`;

exports[`Rate 2.render test renders correctly 2`] = `
<span
  class="kd-rate kd-rate-size-large"
>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
</span>
`;

exports[`Rate 2.render test renders correctly 3`] = `
<span
  class="kd-rate kd-rate-size-middle"
>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
</span>
`;

exports[`Rate 2.render test renders correctly 4`] = `
<span
  class="kd-rate kd-rate-size-small"
>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
  <span
    class="kd-rate-item"
  >
    <div
      class="kd-rate-icon-view"
    >
      <span
        class="kd-rate-icon-first"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
      <span
        class="kd-rate-icon-second"
      >
        <span
          class="kd-rate-icon-wrapper"
        >
          <i
            class="kdicon kdicon-star"
          />
        </span>
      </span>
    </div>
  </span>
</span>
`;

exports[`Rate 4. render null or undefined without errors render null or undefined buttons without errors 1`] = `
<Rate>
  <span
    className="kd-rate kd-rate-size-middle"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="kd-rate-item"
      key="1"
    >
      <div
        className="kd-rate-icon-view"
      >
        <span
          className="kd-rate-icon-first"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <span
            className="kd-rate-icon-wrapper"
          >
            <Icon
              type="star"
            >
              <i
                className="kdicon kdicon-star"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
        </span>
        <span
          className="kd-rate-icon-second"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <span
            className="kd-rate-icon-wrapper"
          >
            <Icon
              type="star"
            >
              <i
                className="kdicon kdicon-star"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
        </span>
      </div>
    </span>
    <span
      className="kd-rate-item"
      key="2"
    >
      <div
        className="kd-rate-icon-view"
      >
        <span
          className="kd-rate-icon-first"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <span
            className="kd-rate-icon-wrapper"
          >
            <Icon
              type="star"
            >
              <i
                className="kdicon kdicon-star"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
        </span>
        <span
          className="kd-rate-icon-second"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <span
            className="kd-rate-icon-wrapper"
          >
            <Icon
              type="star"
            >
              <i
                className="kdicon kdicon-star"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
        </span>
      </div>
    </span>
    <span
      className="kd-rate-item"
      key="3"
    >
      <div
        className="kd-rate-icon-view"
      >
        <span
          className="kd-rate-icon-first"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <span
            className="kd-rate-icon-wrapper"
          >
            <Icon
              type="star"
            >
              <i
                className="kdicon kdicon-star"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
        </span>
        <span
          className="kd-rate-icon-second"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <span
            className="kd-rate-icon-wrapper"
          >
            <Icon
              type="star"
            >
              <i
                className="kdicon kdicon-star"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
        </span>
      </div>
    </span>
    <span
      className="kd-rate-item"
      key="4"
    >
      <div
        className="kd-rate-icon-view"
      >
        <span
          className="kd-rate-icon-first"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <span
            className="kd-rate-icon-wrapper"
          >
            <Icon
              type="star"
            >
              <i
                className="kdicon kdicon-star"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
        </span>
        <span
          className="kd-rate-icon-second"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <span
            className="kd-rate-icon-wrapper"
          >
            <Icon
              type="star"
            >
              <i
                className="kdicon kdicon-star"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
        </span>
      </div>
    </span>
    <span
      className="kd-rate-item"
      key="5"
    >
      <div
        className="kd-rate-icon-view"
      >
        <span
          className="kd-rate-icon-first"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <span
            className="kd-rate-icon-wrapper"
          >
            <Icon
              type="star"
            >
              <i
                className="kdicon kdicon-star"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
        </span>
        <span
          className="kd-rate-icon-second"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <span
            className="kd-rate-icon-wrapper"
          >
            <Icon
              type="star"
            >
              <i
                className="kdicon kdicon-star"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
        </span>
      </div>
    </span>
  </span>
</Rate>
`;
